mozjs/js/rust/src/jsslotconversions.rs

```diff
@@ -0,0 +1,250 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Conversion of Rust values to and from JS Slots
+
+#![deny(missing_docs)]
+
+use conversions::{ConversionResult, ConversionBehavior, FromJSValConvertible,
+                  ToJSValConvertible};
+use jsapi::root::*;
+use jsval;
+use std::mem;
+
+/// A trait to record the number of slots required for the current type
+pub trait NumSlots {
+    /// Number of slots required to store Self in inheritance
+    const NUM_SLOTS: u32 = 1;
+
+    /// Number of slots required to store Self as pointer
+    const NUM_SLOTS_PTR: u32 = 1;
+}
+
+/// A trait to store and retrieve data with Rust type to JS slots
+pub trait ToFromJsSlots : NumSlots {
+    /// get the self from the slots
+    /// panics if it has already been taken out of slots
+    unsafe fn from_slots(object: *mut JSObject, cx: *mut JSContext, offset: u32) -> Self;
```

Comment:
For copy types, we just want to return a copy and it will never panic.

For non-copy types, we probably want to return `Option<Self>` and let the caller decide whether to panic or not by letting them `unwrap()` or match on the return and handle `None`.

Because there are two different behaviors that are both desired, I would usually say this should be two different traits, but because we need a single trait to call from codegen, we can represent this behavior with an associated type on the trait:

```rust
type Target = Self; // defaults to same in/out type

// This is the same
unsafe fn into_slots(self, cx: *mut JSObject, obj: *mut JSObject, offset: u32);

// Because this returns `Self::Target`, we can override the return type as needed
// to be `Option<Self>` etc when appropriate.
unsafe fn from_slots(cx: *mut JSObject, obj: *mut JSObject, offset: u32) -> Self::Target;
```